### DISCUSSION TITLE
Do we need to rewrite everything from scratch...?

### DISCUSSION THREAD

**Discussion by Thesola10** [Thesola10/osu-DIVA#26] [Ideas]
Ever since I wrote #17 and had a good look at osu!classic's code, I've been wondering if we aren't barking up the wrong tree, and need to revisit the entire architecture before it's too late.

For instance, timing code on osu!DIVA being in DrawableDivaHitObject.cs (note the Drawable) is part of the reason for #10, or certainly didn't make it easier to fix.

Far from the idea of speaking ill of @Artemis-chan's work -- quite the opposite, without your project I wouldn't even dare start work on an osu! ruleset...

My proposal is to fork the code from osu!classic, then build osu!DIVA features on top, given that osu!classic is the closest in feature set to this project. This will allow us to inherit a known-working architecture and build on the shoulders of giants.

This would also make it easier to _backport_ code from other stock osu! rulesets, like Mania, for our conversion decision algorithm, or our editor.

**Comment by Thesola10** [comment]
Okay, I'm dramatizing a bit, by definition we could keep code that was already backported, and slot it in place on the new codebase, so not too big an undertaking, plus it would give us a head start on #9

**Comment by Artemis-chan** [comment]
I would definitely agree, as you know I've been just bodging something playble instead of researching the official rulesets for days. But as the project's getting this big I too am seeing quite a lot of stuff getting messy. Like the recent  #24 is quite bodged as well, gotta rewrite quite the whole hitobject thing for the sliders.

**Reply by Thesola10** [reply]
I'll start poking around with `osu.Game.Rulesets.Osu` until it looks like osu!DIVA, then push an orphaned branch, which we will eventually be able to make into the main branch.

**Comment by Thesola10** [comment]
Completely crazy idea: writing osu!DIVA in F#

I've been playing around with Haskell and Elm, both pure functional programming languages, and I feel that something as heavily event- and math-oriented as an osu! ruleset may benefit from at least some components being written in a functional language.

Given that F# is a .NET CIL language, I'd need to research interop between it and C#, and of course this isn't _guaranteed_ to happen anyway.

idk ¯\\\_(ツ)_/¯
